// highcharts-angular/esm2015/lib/highcharts-chart.module.js
import { NgModule } from '@angular/core';
import { HighchartsChartComponent } from './highcharts-chart.component';
import * as ɵngcc0 from '@angular/core';
export class HighchartsChartModule {
}
HighchartsChartModule.ɵfac = function HighchartsChartModule_Factory(t) { return new (t || HighchartsChartModule)(); };
HighchartsChartModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: HighchartsChartModule });
HighchartsChartModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({});
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(HighchartsChartModule, [{
        type: NgModule,
        args: [{
                declarations: [HighchartsChartComponent],
                exports: [HighchartsChartComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(HighchartsChartModule, { declarations: function () { return [HighchartsChartComponent]; }, exports: function () { return [HighchartsChartComponent]; } }); })();
