// ngx-auth/esm5/auth.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * Essential service for authentication
 * @abstract
 */
var /**
 * Essential service for authentication
 * @abstract
 */
AuthService = /** @class */ (function () {
    function AuthService() {
    }
    return AuthService;
}());
/**
 * Essential service for authentication
 * @abstract
 */
export { AuthService };
if (false) {
    /**
     * Check, if user already authorized.
     * Should return Observable with true or false values
     * @abstract
     * @return {?}
     */
    AuthService.prototype.isAuthorized = function () { };
    /**
     * Get access token
     * Should return access token in Observable from e.g.
     * localStorage
     * @abstract
     * @return {?}
     */
    AuthService.prototype.getAccessToken = function () { };
    /**
     * Function, that should perform refresh token
     * Should be successfully completed so interceptor
     * can execute pending requests or retry original one
     * @abstract
     * @return {?}
     */
    AuthService.prototype.refreshToken = function () { };
    /**
     * Function, checks response of failed request to determine,
     * whether token be refreshed or not.
     *
     * Essentially checks status
     * @abstract
     * @param {?} response
     * @param {?=} request
     * @return {?}
     */
    AuthService.prototype.refreshShouldHappen = function (response, request) { };
    /**
     * Verify that outgoing request is refresh-token,
     * so interceptor won't intercept this request
     * @abstract
     * @param {?} request
     * @return {?}
     */
    AuthService.prototype.verifyRefreshToken = function (request) { };
    /**
     * Checks if request must be skipped by interceptor.
     * Useful for requests such as request token which doesn't require token in headers
     * @abstract
     * @param {?} request
     * @return {?}
     */
    AuthService.prototype.skipRequest = function (request) { };
    /**
     * Add token to headers, dependent on server
     * set-up, by default adds a bearer token.
     * Called by interceptor.
     * To change behavior, override this method.
     * @abstract
     * @param {?} token
     * @return {?}
     */
    AuthService.prototype.getHeaders = function (token) { };
    /**
     * Saves last interrupted url inside of the service for further reusage,
     * e.g. restoring interrupted page after logging in
     * @abstract
     * @param {?} url
     * @return {?}
     */
    AuthService.prototype.setInterruptedUrl = function (url) { };
    /**
     * Verify that outgoing request is refresh-token,
     * so interceptor won't intercept this request
     * @deprecated Due to illogical meaning/functionality this method is deprecated
     * @see verifyRefreshToken
     * @abstract
     * @param {?} url
     * @return {?}
     */
    AuthService.prototype.verifyTokenRequest = function (url) { };
}
