// ng-pick-datetime-ex/esm2020/lib/date-time/adapter/dayjs-adapter/dayjs-date-time.module.mjs
/**
 * dayjs-date-time.module
 */
import { NgModule } from '@angular/core';
import { DayjsDateTimeAdapter, OWL_DAYJS_DATE_TIME_ADAPTER_OPTIONS } from './dayjs-date-time-adapter.class';
import { OWL_DAYJS_DATE_TIME_FORMATS } from './dayjs-date-time-format.class';
import { DateTimeAdapter, OWL_DATE_TIME_LOCALE } from '../date-time-adapter.class';
import { OWL_DATE_TIME_FORMATS } from '../date-time-format.class';
import * as i0 from "@angular/core";
// import { DateTimeAdapter, OWL_DATE_TIME_FORMATS, OWL_DATE_TIME_LOCALE_PROVIDER } from 'ng-pick-datetime';
export class DayjsDateTimeModule {
}
DayjsDateTimeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: DayjsDateTimeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DayjsDateTimeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: DayjsDateTimeModule });
DayjsDateTimeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: DayjsDateTimeModule, providers: [
        {
            provide: DateTimeAdapter,
            useClass: DayjsDateTimeAdapter,
            deps: [OWL_DATE_TIME_LOCALE, OWL_DAYJS_DATE_TIME_ADAPTER_OPTIONS]
        },
    ] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: DayjsDateTimeModule, decorators: [{
            type: NgModule,
            args: [{
                    providers: [
                        {
                            provide: DateTimeAdapter,
                            useClass: DayjsDateTimeAdapter,
                            deps: [OWL_DATE_TIME_LOCALE, OWL_DAYJS_DATE_TIME_ADAPTER_OPTIONS]
                        },
                    ],
                }]
        }] });
export class OwlDayjsDateTimeModule {
}
OwlDayjsDateTimeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: OwlDayjsDateTimeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
OwlDayjsDateTimeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: OwlDayjsDateTimeModule, imports: [DayjsDateTimeModule] });
OwlDayjsDateTimeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: OwlDayjsDateTimeModule, providers: [{ provide: OWL_DATE_TIME_FORMATS, useValue: OWL_DAYJS_DATE_TIME_FORMATS }], imports: [[DayjsDateTimeModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: OwlDayjsDateTimeModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [DayjsDateTimeModule],
                    providers: [{ provide: OWL_DATE_TIME_FORMATS, useValue: OWL_DAYJS_DATE_TIME_FORMATS }],
                }]
        }] });
