// ngx-viewer/esm2015/lib/ngx-viewer.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { NgxViewerDirective } from './ngx-viewer.directive';
import * as ɵngcc0 from '@angular/core';
export class NgxViewerModule {
}
NgxViewerModule.ɵfac = function NgxViewerModule_Factory(t) { return new (t || NgxViewerModule)(); };
NgxViewerModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: NgxViewerModule });
NgxViewerModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NgxViewerModule, [{
        type: NgModule,
        args: [{
                imports: [],
                declarations: [NgxViewerDirective],
                exports: [NgxViewerDirective]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgxViewerModule, { declarations: function () { return [NgxViewerDirective]; }, exports: function () { return [NgxViewerDirective]; } }); })();
