// @jonijnm/amazing-time-picker/esm2015/lib/atp-time-picker/atp-time-picker.component.js
import { ApplicationRef, Component, ComponentFactoryResolver, EventEmitter, Output, ViewChild, ViewContainerRef } from '@angular/core';
import { TimePickerComponent } from '../time-picker/time-picker.component';
import { Subject } from 'rxjs';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["container"];
export class AtpTimePickerComponent {
    constructor(resolver, appRef) {
        this.resolver = resolver;
        this.appRef = appRef;
        this.timeSelected = new EventEmitter();
        this.config = {};
    }
    ngOnInit() {
        let config = this.config;
        config = {
            time: config.time || '00:00',
            theme: [
                'light',
                'dark',
                'material'
            ].indexOf(config.theme) > 0 ? config.theme : 'light' || config.theme || 'light',
            rangeTime: config.rangeTime || { start: '0:0', end: '24:0' },
            arrowStyle: config.arrowStyle || {}
        };
        config.arrowStyle = {
            background: (config.arrowStyle.background) ?
                config.arrowStyle.background : config.theme !== undefined ?
                config.theme === 'dark' ? 'rgb(128, 203, 196)' : 'blue' : 'blue',
            color: config.arrowStyle.color || '#fff'
        };
        const cfr = this.resolver.resolveComponentFactory(TimePickerComponent);
        const tsc = this.container.createComponent(cfr);
        tsc.instance.subject = new Subject();
        tsc.instance._ref = tsc;
        tsc.instance.appRef = this.appRef;
        tsc.instance.timerElement = '';
        tsc.instance.config = config;
        tsc.instance.activeModal = true;
        tsc.instance.isPopup = false;
        tsc.instance.ParseStringToTime(config.time);
        tsc.instance.subject.asObservable().subscribe(time => {
            this.timeSelected.emit(time);
        });
    }
}
AtpTimePickerComponent.ɵfac = function AtpTimePickerComponent_Factory(t) { return new (t || AtpTimePickerComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ComponentFactoryResolver), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ApplicationRef)); };
AtpTimePickerComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: AtpTimePickerComponent, selectors: [["atp-time-picker"]], viewQuery: function AtpTimePickerComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 5, ViewContainerRef);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.container = _t.first);
    } }, outputs: { timeSelected: "timeSelected" }, decls: 2, vars: 0, consts: [["container", ""]], template: function AtpTimePickerComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelement(0, "div", null, 0);
    } }, styles: [".atp-time-picker[_ngcontent-%COMP%]   .icon-container[_ngcontent-%COMP%]{display:inline-block;margin-right:.2em}.atp-time-picker[_ngcontent-%COMP%]   .icon-container[_ngcontent-%COMP%]   svg[_ngcontent-%COMP%]{cursor:pointer;position:relative;top:.5em}.atp-time-picker[_ngcontent-%COMP%]   .icon-container[_ngcontent-%COMP%]     i{cursor:pointer}"] });
AtpTimePickerComponent.ctorParameters = () => [
    { type: ComponentFactoryResolver },
    { type: ApplicationRef }
];
AtpTimePickerComponent.propDecorators = {
    container: [{ type: ViewChild, args: ['container', { read: ViewContainerRef },] }],
    timeSelected: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AtpTimePickerComponent, [{
        type: Component,
        args: [{
                selector: 'atp-time-picker',
                template: "<div #container></div>",
                styles: [".atp-time-picker .icon-container{display:inline-block;margin-right:.2em}.atp-time-picker .icon-container svg{cursor:pointer;position:relative;top:.5em}.atp-time-picker .icon-container ::ng-deep i{cursor:pointer}\n"]
            }]
    }], function () { return [{ type: ɵngcc0.ComponentFactoryResolver }, { type: ɵngcc0.ApplicationRef }]; }, { timeSelected: [{
            type: Output
        }], container: [{
            type: ViewChild,
            args: ['container', { read: ViewContainerRef }]
        }] }); })();
