// @hallysonh/ngx-imageviewer/esm2015/lib/imagecache.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * @record
 */
import * as ɵngcc0 from '@angular/core';
export function CacheDef() { }
function CacheDef_tsickle_Closure_declarations() {
    /** @type {?} */
    CacheDef.prototype.url;
    /** @type {?} */
    CacheDef.prototype.page;
    /** @type {?} */
    CacheDef.prototype.image;
}
export class ImageCacheService {
    constructor() {
        this._cache = [];
    }
    /**
     * @return {?}
     */
    get cache() {
        return this._cache;
    }
    /**
     * @param {?} url
     * @param {?} page
     * @return {?}
     */
    getCache(url, page) {
        return this.cache.find(i => i.url === url && i.page === page);
    }
    /**
     * @param {?} url
     * @param {?} page
     * @return {?}
     */
    getImage(url, page) {
        const /** @type {?} */ c = this.getCache(url, page);
        return c ? c.image : null;
    }
    /**
     * @param {?} url
     * @param {?} page
     * @param {?} image
     * @return {?}
     */
    saveImage(url, page, image) {
        const /** @type {?} */ cache = this.getCache(url, page);
        if (cache) {
            cache.image = image;
        }
        else {
            this.cache.push({ url, page, image });
        }
    }
    /**
     * @return {?}
     */
    disposeCache() {
        this.cache.forEach(i => URL.revokeObjectURL(i.image.src));
        this._cache = [];
    }
}
ImageCacheService.ɵfac = function ImageCacheService_Factory(t) { return new (t || ImageCacheService)(); };
ImageCacheService.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: ImageCacheService, factory: ImageCacheService.ɵfac, providedIn: 'root' });
/** @nocollapse */
ImageCacheService.ctorParameters = () => [];
/** @nocollapse */ ImageCacheService.ngInjectableDef = i0.defineInjectable({ factory: function ImageCacheService_Factory() { return new ImageCacheService(); }, token: ImageCacheService, providedIn: "root" });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ImageCacheService, [{
        type: Injectable,
        args: [{ providedIn: 'root' }]
    }], function () { return []; }, null); })();
function ImageCacheService_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    ImageCacheService.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    ImageCacheService.ctorParameters;
    /** @type {?} */
    ImageCacheService.prototype._cache;
}
