// @hallysonh/ngx-imageviewer/esm5/lib/imagecache.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * @record
 */
import * as ɵngcc0 from '@angular/core';
export function CacheDef() { }
function CacheDef_tsickle_Closure_declarations() {
    /** @type {?} */
    CacheDef.prototype.url;
    /** @type {?} */
    CacheDef.prototype.page;
    /** @type {?} */
    CacheDef.prototype.image;
}
var ImageCacheService = /** @class */ (function () {
    function ImageCacheService() {
        this._cache = [];
    }
    Object.defineProperty(ImageCacheService.prototype, "cache", {
        get: /**
         * @return {?}
         */
        function () {
            return this._cache;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} url
     * @param {?} page
     * @return {?}
     */
    ImageCacheService.prototype.getCache = /**
     * @param {?} url
     * @param {?} page
     * @return {?}
     */
    function (url, page) {
        return this.cache.find(function (i) { return i.url === url && i.page === page; });
    };
    /**
     * @param {?} url
     * @param {?} page
     * @return {?}
     */
    ImageCacheService.prototype.getImage = /**
     * @param {?} url
     * @param {?} page
     * @return {?}
     */
    function (url, page) {
        var /** @type {?} */ c = this.getCache(url, page);
        return c ? c.image : null;
    };
    /**
     * @param {?} url
     * @param {?} page
     * @param {?} image
     * @return {?}
     */
    ImageCacheService.prototype.saveImage = /**
     * @param {?} url
     * @param {?} page
     * @param {?} image
     * @return {?}
     */
    function (url, page, image) {
        var /** @type {?} */ cache = this.getCache(url, page);
        if (cache) {
            cache.image = image;
        }
        else {
            this.cache.push({ url: url, page: page, image: image });
        }
    };
    /**
     * @return {?}
     */
    ImageCacheService.prototype.disposeCache = /**
     * @return {?}
     */
    function () {
        this.cache.forEach(function (i) { return URL.revokeObjectURL(i.image.src); });
        this._cache = [];
    };
    /** @nocollapse */
    ImageCacheService.ctorParameters = function () { return []; };
    /** @nocollapse */ ImageCacheService.ngInjectableDef = i0.defineInjectable({ factory: function ImageCacheService_Factory() { return new ImageCacheService(); }, token: ImageCacheService, providedIn: "root" });
ImageCacheService.ɵfac = function ImageCacheService_Factory(t) { return new (t || ImageCacheService)(); };
ImageCacheService.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: ImageCacheService, factory: function (t) { return ImageCacheService.ɵfac(t); }, providedIn: 'root' });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ImageCacheService, [{
        type: Injectable,
        args: [{ providedIn: 'root' }]
    }], function () { return []; }, null); })();
    return ImageCacheService;
}());
export { ImageCacheService };
function ImageCacheService_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    ImageCacheService.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    ImageCacheService.ctorParameters;
    /** @type {?} */
    ImageCacheService.prototype._cache;
}
