// @jonijnm/amazing-time-picker/esm2015/lib/amazing-time-picker.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TimePickerComponent } from './time-picker/time-picker.component';
import { AtpTimePickerComponent } from './atp-time-picker/atp-time-picker.component';
import { AtpDirective } from './atp.directive';
import { AmazingTimePickerService } from './amazing-time-picker.service';
import { AtpCoreService } from './atp-core.service';
import * as ɵngcc0 from '@angular/core';
export class AmazingTimePickerModule {
}
AmazingTimePickerModule.ɵfac = function AmazingTimePickerModule_Factory(t) { return new (t || AmazingTimePickerModule)(); };
AmazingTimePickerModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: AmazingTimePickerModule });
AmazingTimePickerModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        AmazingTimePickerService,
        AtpCoreService
    ], imports: [[
            CommonModule,
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AmazingTimePickerModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                ],
                declarations: [
                    TimePickerComponent,
                    AtpTimePickerComponent,
                    AtpDirective
                ],
                providers: [
                    AmazingTimePickerService,
                    AtpCoreService
                ],
                entryComponents: [
                    TimePickerComponent
                ],
                exports: [
                    TimePickerComponent,
                    AtpTimePickerComponent,
                    AtpDirective
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AmazingTimePickerModule, { declarations: function () { return [TimePickerComponent,
        AtpTimePickerComponent,
        AtpDirective]; }, imports: function () { return [CommonModule]; }, exports: function () { return [TimePickerComponent,
        AtpTimePickerComponent,
        AtpDirective]; } }); })();
