// @jonijnm/amazing-time-picker/esm2015/lib/atp.directive.js
import { Directive, EventEmitter, HostListener, Output, ViewContainerRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { AmazingTimePickerService } from './amazing-time-picker.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './amazing-time-picker.service';
export class AtpDirective {
    constructor(viewContainerRef, atp) {
        this.viewContainerRef = viewContainerRef;
        this.atp = atp;
        this.myClick = new EventEmitter();
        this.onChange = (x) => {
        };
        this.elementRef = this.viewContainerRef.element;
    }
    onClick(e) {
        const ele = this.viewContainerRef.element.nativeElement;
        const time = ele.value;
        const theme = ele.getAttribute('theme');
        const start = ele.getAttribute('start');
        const end = ele.getAttribute('end');
        const locale = ele.getAttribute('locale') || 'en';
        const changeToMinutes = ele.getAttribute('changeToMinutes') === 'true';
        const animation = ele.getAttribute('animation');
        const preference = ele.getAttribute('preference') || null;
        const onlyHour = ele.getAttribute('onlyHour') === 'true';
        const onlyMinute = ele.getAttribute('onlyMinute') === 'true';
        const onlyAM = ele.getAttribute('onlyAM') === 'true';
        const onlyPM = ele.getAttribute('onlyPM') === 'true';
        let arrowStyle = ele.getAttribute('arrowStyle');
        arrowStyle = (arrowStyle) ? JSON.parse(arrowStyle.replace(new RegExp('\'', 'g'), '"')) : '';
        const timePickerFunction = this.atp.open({
            time,
            theme,
            rangeTime: { start, end },
            'arrowStyle': arrowStyle,
            locale,
            changeToMinutes,
            animation,
            onlyHour,
            onlyMinute,
            onlyAM,
            onlyPM,
            preference
        });
        timePickerFunction.afterClose().subscribe(retTime => {
            this.writeValue(retTime); // update the native element
            this.onChange(retTime); // update the form value (if there's a form)
        });
    }
    onInput(e) {
        this.onChange(e.srcElement.value);
    }
    writeValue(value) {
        if (this.elementRef) {
            this.elementRef.nativeElement.value = value;
        }
    }
    registerOnChange(fn) {
        this.onChange = fn;
    }
    registerOnTouched(fn) {
    }
}
AtpDirective.ɵfac = function AtpDirective_Factory(t) { return new (t || AtpDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ViewContainerRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.AmazingTimePickerService)); };
AtpDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: AtpDirective, selectors: [["input", "atp-time-picker", ""]], hostBindings: function AtpDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function AtpDirective_click_HostBindingHandler($event) { return ctx.onClick($event); })("input", function AtpDirective_input_HostBindingHandler($event) { return ctx.onInput($event); });
    } }, outputs: { myClick: "myClick" }, features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: NG_VALUE_ACCESSOR,
                useExisting: AtpDirective,
                multi: true
            }
        ])] });
AtpDirective.ctorParameters = () => [
    { type: ViewContainerRef },
    { type: AmazingTimePickerService }
];
AtpDirective.propDecorators = {
    myClick: [{ type: Output }],
    onClick: [{ type: HostListener, args: ['click', ['$event'],] }],
    onInput: [{ type: HostListener, args: ['input', ['$event'],] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AtpDirective, [{
        type: Directive,
        args: [{
                selector: 'input[atp-time-picker]',
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: AtpDirective,
                        multi: true
                    }
                ]
            }]
    }], function () { return [{ type: ɵngcc0.ViewContainerRef }, { type: ɵngcc1.AmazingTimePickerService }]; }, { myClick: [{
            type: Output
        }], onClick: [{
            type: HostListener,
            args: ['click', ['$event']]
        }], onInput: [{
            type: HostListener,
            args: ['input', ['$event']]
        }] }); })();
