// @jonijnm/amazing-time-picker/esm2015/lib/preferences.js
const arabic = new Intl.NumberFormat('ar-AE');
const persian = new Intl.NumberFormat('fa-IR');
const ɵ0 = (x) => persian.format(x), ɵ1 = (x) => {
    let exp = persian.format(x);
    if (exp.length === 1) {
        exp = persian.format(0) + exp;
    }
    return exp;
}, ɵ2 = (x) => x === 'AM' ? 'صبح' : 'عصر', ɵ3 = (x) => persian.format(x), ɵ4 = (x) => persian.format(x);
export const PersianPreference = {
    hour: ɵ0,
    minute: ɵ1,
    separator: ':',
    period: ɵ2,
    clockHour: ɵ3,
    clockMinute: ɵ4,
    labels: {
        ok: 'تایید',
        cancel: 'لغو'
    }
};
const ɵ5 = (x) => arabic.format(x), ɵ6 = (x) => {
    let exp = arabic.format(x);
    if (exp.length === 1) {
        exp = arabic.format(0) + exp;
    }
    return exp;
}, ɵ7 = (x) => x === 'AM' ? 'صباحا' : 'مساء', ɵ8 = (x) => arabic.format(x), ɵ9 = (x) => arabic.format(x);
export const ArabicPreference = {
    hour: ɵ5,
    minute: ɵ6,
    separator: ':',
    period: ɵ7,
    clockHour: ɵ8,
    clockMinute: ɵ9,
    labels: {
        ok: 'حسنا',
        cancel: 'إلغاء'
    }
};
const ɵ10 = (x) => x, ɵ11 = (x) => {
    let exp = x;
    if (exp.length === 1) {
        exp = '۰' + exp;
    }
    return exp;
}, ɵ12 = (x) => x === 'AM' ? '上午' : '下午', ɵ13 = (x) => x, ɵ14 = (x) => x;
export const ChinesePreference = {
    hour: ɵ10,
    minute: ɵ11,
    separator: ':',
    period: ɵ12,
    clockHour: ɵ13,
    clockMinute: ɵ14,
    labels: {
        ok: '确定',
        cancel: '取消'
    }
};
export const Preference = (locale) => {
    switch (locale) {
        case 'fa':
            return PersianPreference;
        case 'ar':
            return ArabicPreference;
        case 'zh':
            return ChinesePreference;
        default:
            return null;
    }
};
export { ɵ0, ɵ1, ɵ2, ɵ3, ɵ4, ɵ5, ɵ6, ɵ7, ɵ8, ɵ9, ɵ10, ɵ11, ɵ12, ɵ13, ɵ14 };
